contract_comm/evm.go

```diff
@@ -67,6 +67,11 @@ func NewEVMContext(msg types.Message, header *types.Header, chain ChainContext,
 		beneficiary = *author
 	}
 
+	var engine consensus.Engine
+	if chain != nil {
```

Comment:
I have two comments regarding this nil check:

1) Be aware that you are checking if the interface value is nil, NOT if the underlying value of the interface is nil.  So if this function is ever invoked with a `chain` parameter that is set to nil, then the next line will panic (as opposed to if the function is invoked with a "constant" `nil` parameter for that parameter).
2) When would chain be nil?  Seems like if it is, then your new `getEpochSize` precompile will panic, as it assumes that engine will be non nil.

Reply:
section 7.5.1 of the golang book has a good description about this.

In summary, the way to solve it is that to make sure that the arguments passed to the function is an interface type, as opposed to a "real" type.

Reply:
I don't have the book handy, how do I do this?

Reply:
IMO, I think it's best to make it so that the concrete value of `chain` is always non-nil and to not do the `if` check.

If that's not doable, then it will get messy.  E.g. you could use introspective via the `reflect` package, and do something like this:

`if c == nil || (reflect.ValueOf(c).Kind() == reflect.Ptr && reflect.ValueOf(c).IsNil())`

where `c` is the interface variable.

Reply:
IMO, I think it would be best to always assume that chain is not `nil` and it's contract value is not nil, so that the `if` condition is removed.   Seems like we have precompiles that now depend on it not being nil, such as your `epochsize` and the validator set related ones.

Reply:
Still not totally clear on this, let's chat in person

Reply:
Removed this check